ddtrace/settings/_inferred_service.py

```diff
@@ -0,0 +1,199 @@
+
+        return ServiceMetadata(""), False
+
+    def abs_path(self, a, wd):
+        # Converts the given path to an absolute path
+        return pathlib.Path(wd) / a
+
+    # deduce_package_name is walking until a `__init__.py` is not found.
+    # All the dir traversed are joined then with `.`
+    def deduce_package_name(self, fp):
+        up = pathlib.Path(fp).parent
+        current = fp
+        traversed = []
+
+        while current != up:
+            if not pathlib.Path(current, INIT_PY).exists():
+                break
+            traversed.insert(0, pathlib.Path(current).name)
+            current = up
+            up = pathlib.Path(current).parent
+
+        return ".".join(traversed), len(traversed) > 0
+
+    # findNearestTopLevel returns the top level dir containing a .py file starting walking up from fp
+    def find_nearest_top_level(self, fp):
+        up = pathlib.Path(fp).parent
+        current = fp
+        last = current
+
+        while current != up:
+            if not fnmatch.filter(os.listdir(current), ALL_PY_FILES):
+                break
+            last = current
+            current = up
+            up = pathlib.Path(current).parent
+
+        return pathlib.Path(last).name
+
+
+class GunicornDetector:
+    def __init__(self, ctx):
+        self.ctx = ctx
+        name = "gunicorn"
+        self.name = name
+        self.pattern = name
+
+    def expected_command_name(self):
+        return "gunicorn"
+
+    def detect(self, args):
+        from_env = os.getenv(GUNICORN_CMD_ARGS)
+        if from_env:
+            name, ok = self.extract_gunicorn_name_from(from_env.split())
+            if ok:
+                return ServiceMetadata(name), True
+
+        wsgi_app = os.getenv(WSGI_APP_ENV)
+        if wsgi_app:
+            return ServiceMetadata(self.parse_name_from_wsgi_app(wsgi_app)), True
+
+        name, ok = self.extract_gunicorn_name_from(args)
+        if ok:
+            return ServiceMetadata(name), True
+
+        return ServiceMetadata("gunicorn"), True
+
+    def extract_gunicorn_name_from(self, args):
+        skip = False
+        capture = False
+
+        for arg in args:
+            if capture:
+                return arg, True
+            if skip:
+                skip = False
+                continue
+            if arg.startswith("-"):
+                if arg == "-n":
+                    capture = True
+                    continue
+                skip = "=" not in arg
+                if skip:
+                    continue
+                if arg.startswith("--name="):
+                    return arg[len("--name=") :], True
+            else:
+                return self.parse_name_from_wsgi_app(args[-1]), True
+        return "", False
+
+    def parse_name_from_wsgi_app(self, wsgi_app):
+        name, _, _ = wsgi_app.partition(":")
+        return name
+
+
+def detect_service(args, detector_classes=[PythonDetector, GunicornDetector]):
```

Comment:
You should cache the result (unless we think it should be dynamic)

Reply:
done.